src/syrupy/extensions/base.py

```diff
@@ -214,36 +223,140 @@ def __ensure_snapshot_dir(self, *, index: int) -> None:
+                else:
+                    should_unstage = (
+                        staged_diffed_line.is_complete
+                        or (staged_diffed_line.has_snapshot and is_snapshot_line)
+                        or (staged_diffed_line.has_received and is_received_line)
+                        or (staged_diffed_line.is_context and not is_context_line)
+                    )
+                    if should_unstage:
+                        yield staged_diffed_line
+                        staged_diffed_line = None
+                    elif is_snapshot_line:
+                        staged_diffed_line.a = line
+                    elif is_received_line:
+                        staged_diffed_line.b = line
+                    elif is_context_line:
+                        staged_diffed_line.c.append(line)
+
+            if not staged_diffed_line:
+                if is_diff_line:
+                    raise RuntimeWarning(
+                        "Encounted a diff line without any previously staged line"
+                    )
+                elif is_snapshot_line:
+                    staged_diffed_line = DiffedLine(a=line)
+                elif is_received_line:
+                    staged_diffed_line = DiffedLine(b=line)
+                elif is_context_line:
+                    staged_diffed_line = DiffedLine(c=[line])
+
+        if staged_diffed_line:
+            yield staged_diffed_line
+
+    def __format_line(
+        self,
+        line: str,
+        diff_markers: str,
+        line_style: Callable[[str], str],
+        diff_style: Callable[[str], str],
+        show_ends: bool,
     ) -> str:
+        if show_ends:
+            for old, new in self._ends.items():
+                line = line.replace(old, mute(new))
+        else:
+            line = self.__strip_ends(line)
         return "".join(
-            emphasize(line_style(char)) if str(marker) in "-+^" else line_style(char)
-            for marker, char in zip_longest(marker_line.strip(), line)
+            diff_style(char) if str(marker) in "-+^" else line_style(char)
+            for marker, char in zip_longest(diff_markers.rstrip(), line)
         )
 
+    def __limit_context(self, lines: List[str]) -> Iterator[str]:
+        yield from lines[: self._context_line_count]
+        num_lines = len(lines)
+        if num_lines:
+            if num_lines > self._context_line_max:
+                count_leading_whitespace: Callable[[str], int] = (
+                    lambda s: len(s) - len(s.lstrip())  # noqa: E731
+                )
+                if self._context_line_count:
+                    num_space = (
+                        count_leading_whitespace(lines[self._context_line_count - 1])
+                        + count_leading_whitespace(lines[-self._context_line_count])
+                    ) // 2
+                else:
+                    num_space = count_leading_whitespace(lines[num_lines // 2])
```

Comment:
Simpler to reason about but leaves room for `...` having weird indentation
```suggestion
                mid_line = lines[num_lines // 2]
                num_space = len(mid_line) - len(mid_line.lstrip())
```